### ISSUE TITLE
"fwrite(): send of 4 bytes failed with errno=32 Broken pipe at \/var\/www\/owncloud\/apps\/files_antivirus\/lib\/scanner.php#68

### ISSUE THREAD

**Issue by Daniae** [owncloud/files_antivirus#69]
Hi,

i have a Problem with ownCloud 8.1.0 and files_antivirus 0.7.0.1. I got the following message several times in a second in the owncloud.log:

remoteAddr":"","app":"PHP","message":"fwrite(): send of 4 bytes failed with errno=32 Broken pipe at \/var\/www\/owncloud\/apps\/files_antivirus\/lib\/scanner.php#68","level":3,

I can not find the source of this problem. Is there anyone who knows why this message appears and how i can fix this problem?

**Comment by DeepDiver1975** [comment]
@VicDeo please have a look - THX

**Comment by Daniae** [comment]
Hi,

i found the cause of this message. The Problem appears if the cron.php from owncloud runs the first time on the server. If i start the Server without using the cron.php, i get no entry in the owncloud.log. I use the Cron entry described in the ownlcoud 8.1 documentation: https://doc.owncloud.org/server/8.1/admin_manual/configuration_server/background_jobs_configuration.html?highlight=cron

I hope these information help you find the source of the Problem. Thank you for the help!

**Comment by VicDeo** [comment]
@Daniae 
What exact cron mode you are using (webcron, ajax. etc)?
What are the files_antivirus app settings?

**Comment by Daniae** [comment]
Hi VicDeo,

i use cron with the cron.php. My Crontab has this entry: 

*/15  *  *  *  * php -f /var/www/owncloud/cron.php > /dev/null 2>&1

This are my settings in the administration panel for the files_antivirus app:

Modus: Deamon Socket
Socket: /var/run/clamav/clamd.ctl
Übertragunsgslänge: 10485760
Aktion: Loggen

This is my /etc/clamav/clamd.conf:

#Automatically Generated by clamav-daemon postinst
#To reconfigure clamd run #dpkg-reconfigure clamav-daemon
#Please read /usr/share/doc/clamav-daemon/README.Debian.gz for details
LocalSocket /var/run/clamav/clamd.ctl
FixStaleSocket true
LocalSocketGroup clamav
LocalSocketMode 666
# TemporaryDirectory is not set to its default /tmp here to make overriding
# the default with environment variables TMPDIR/TMP/TEMP possible
User clamav
AllowSupplementaryGroups false
ScanMail true
ScanArchive true
ArchiveBlockEncrypted false
MaxDirectoryRecursion 15
FollowDirectorySymlinks false
FollowFileSymlinks false
ReadTimeout 180
MaxThreads 12
MaxConnectionQueueLength 15
LogSyslog false
LogRotate true
LogFacility LOG_LOCAL6
LogClean false
LogVerbose true
PidFile /var/run/clamav/clamd.pid
DatabaseDirectory /var/lib/clamav
OfficialDatabaseOnly false
SelfCheck 3600
Foreground false
Debug false
ScanPE true
MaxEmbeddedPE 10M
ScanOLE2 true
ScanPDF true
ScanHTML true
MaxHTMLNormalize 10M
MaxHTMLNoTags 2M
MaxScriptNormalize 5M
MaxZipTypeRcg 1M
ScanSWF true
DetectBrokenExecutables false
ExitOnOOM false
LeaveTemporaryFiles false
AlgorithmicDetection true
ScanELF true
IdleTimeout 30
CrossFilesystems true
PhishingSignatures true
PhishingScanURLs true
PhishingAlwaysBlockSSLMismatch false
PhishingAlwaysBlockCloak false
PartitionIntersection false
DetectPUA false
ScanPartialMessages false
HeuristicScanPrecedence false
StructuredDataDetection false
CommandReadTimeout 5
SendBufTimeout 200
MaxQueue 100
ExtendedDetectionInfo true
OLE2BlockMacros false
ScanOnAccess false
AllowAllMatchScan true
ForceToDisk false
DisableCertCheck false
DisableCache false
MaxScanSize 100M
MaxFileSize 25M
MaxRecursion 10
MaxFiles 10000
MaxPartitions 50
MaxIconsPE 100
StatsEnabled false
StatsPEDisabled true
StatsHostID auto
StatsTimeout 10
StreamMaxLength 25M
LogFile /var/log/clamav/clamav.log
LogTime true
LogFileUnlock false
LogFileMaxSize 0
Bytecode true
BytecodeSecurity TrustSigned
BytecodeTimeout 60000


Do you need more information?

**Comment by mislav-eu** [comment]
I had exactly the same problem with clamav, owncloud.log after few hours was around 10-20GB. 

Error in owncloud.log:
{"reqId":"xy","remoteAddr":"","app":"PHP","message":"fwrite(): send of 4 bytes failed with errno=32 Broken pipe at \/var\/www\/owncloud\/apps\/files_antivirus\/lib\/scanner.php#68","level":3,"time":"2015-08-31T06:40:13+00:00"}

Cron was running as cron (clamav as daemon (socket)):
# crontab -l -u www-data
*/15  *  *  *  * php -f /var/www/owncloud/cron.php > /dev/null 2>&1

At first, I didn't know what is going on, but my disk space used was decreasing kinda fast and nobody was uploading. So I went to troubleshoot a bit, I saw owncloud.log ~ 23GB and this errors were popping up every second, few MBs per sec.

What I've done/tried:
- disable app - no success
- delete folder from apps directory - no success
- apt-get purge clamav clamav-daemon - no success

After that, I've ended up here and commenting out crontab line and rebooting machine. Afterwards owncloud.log didn't get filled with error lines anymore.

That being said, after troubleshooting more deeply, I've found that in .htaccess in owncloud root directory, values upload_max_filesize and & post_max_size were defined with bigger value then server itself has memory assigned.

After lowering these two values, enabling cronjob again, cron was executed successfully without any problems. Please note that I didn't make any changes to clamd.conf.

Maybe it helps!

**Comment by VicDeo** [comment]
@Daniae how much memory do you have on this server?

**Comment by Daniae** [comment]
Hi,

thanks @mislav-eu i tried your solution but for my case it is not working. My system has 8GB Memory. I tried to use 6Gb and only 512MB for the upload_max_filesize and post_max_size but i got still the error also after rebooting the system. Do you need more information?

**Comment by fsbruva** [comment]
Good afternoon-

I am having the same issue. In clamd.conf, my StreamMaxLength is 2000M, my .httaccess file has upload/file size at 512MB, and the system RAM is 16 GB. 

When I attempt to run the cronjob (identical to @Daniae 's config), I get the same error. If I turned on  Debug and Verbose logging in ClamAV, and got the following when I ran the cronjob:
```
got command INSTREAM (9, 14), argument:
Receive thread: INSTREAM: /tmp//clamav-792e5778564386ca029f05178f41a7e2.tmp fd 11
Breaking command loop, mode is no longer MODE_COMMAND
Moved partial command: 680
mode == MODE_STREAM
Got chunksize: 0
Chunks complete
```

So, it seems there is something amiss with how the background scanner communicates chunk sizes. Or, is your [comment](https://github.com/owncloud/files_antivirus/issues/63#issuecomment-137967070) to blame?

**Comment by VicDeo** [comment]
@fsbruva how much free space do you have on partition where `/tmp` directory located?

**Comment by VicDeo** [comment]
Please also post your `Antivirus Configuration` section content at the admin page. Especially value for 
`Stream Length` input

**Comment by DeepDiver1975** [comment]
@cmonteroluque @karlitschek we didn't have a chance to reproduce this issue -> 9.0 :question:

**Comment by karlitschek** [comment]
yes! :-)

**Comment by fsbruva** [comment]
Good morning-

@VicDeo The partition where /tmp is located has 1.8T of free space. 

Additionally, here is the antivirus config:
ModeExecutable: Daemon Daemon (Socket) 
Socket: /var/run/clamd/clamd.sock
Stream Length bytes: 10485760
Action for infected files found while scanning: Only log 

I originally installed the app from the webstore. I disabled and uninstalled the app. When I re-installed it, the Stream Length was back to the default of 1024.

As a side note, I believe I may have determined the issue (in my case). For various reasons, I had to disguise a file. I renamed a .zip file (containing an executable) to .docx. I believe this was the thing that completely borked the antivirus scan. I have since renamed the file back to .zip, and the cronjob completed successfully. I will change the name back to .docx to see if that recreates the error.

**Comment by fsbruva** [comment]
@VicDeo I figured out my issue. My combination of Freebsd, php 5.6, clamav, etc. totally lost its mind when the background scanner attempted to scan an empty text file. I can repeat the above referenced poor behavior by creating a text file using the "Files" tab of the webgui. If the cron job runs, it will hang, and the log will be filled with the above errors. Killing the cron.php process is the only halt. Deleting the offending empty file allows the cron job to run successfully on the next iteration.

Hope this helps someone else!

**Comment by michag86** [comment]
@VicDeo @fsbruva I can confirm this. We had some 0 byte files and I've got the same error message.

After deleting all 0 byte files everything is working fine.

I've created/uploaded some new 0 byte text files but cron.php is still running without problems. How long does it take until these files are scanned in the background?

**Comment by VicDeo** [comment]
@fsbruva @michag86 can you please  test  https://github.com/owncloud/files_antivirus/pull/82 ?

**Comment by fsbruva** [comment]
@michag86 It took my server a few runs (<10?) of cron to index the new files and jam up the background scanner. The cronjob only scans 5 files at a time.

@vicdeo I will be able to test this tomorrow.

**Comment by michag86** [comment]
I cannot reproduce it at the moment. I will wait for the error and apply the fix.

**Comment by fsbruva** [comment]
Sorry for the multiple references - n00b git mistake.... over... and over....

**Comment by felixboehm** [comment]
Please Backport
@karlitschek 

00004215

**Comment by karlitschek** [comment]
@vicdeo what do you think about a backport?

**Comment by VicDeo** [comment]
@karlitschek backport is possible.
@felixboehm is 8.1 and 8.2 enough?

**Comment by felixboehm** [comment]
I think 8.1 and 8.2 is enough. At least a good reason to update.
Thanks

**Comment by karlitschek** [comment]
O.K. Please backport

**Comment by VicDeo** [comment]
#84 
#85

**Comment by CaptionF** [comment]
@VicDeo is it possible to backport to 8.0?

**Comment by VicDeo** [comment]
@CaptionF Are you suffering from this issue?

**Comment by CaptionF** [comment]
customer not ready for upgrade to 8.1 or 8.2!
yes please.

**Comment by CaptionF** [comment]
@VicDeo how long would this take to get a patch for 8.0.x?

**Comment by VicDeo** [comment]
@CaptionF see #86

**Comment by CaptionF** [comment]
@VicDeo Great !

**Comment by VicDeo** [comment]
If I'm not mistaken,  a max size limit should be introduced to fix this completely. 
And it's tricky. :\
There is a dozen of limits and the are not pretty straightforward 
see https://github.com/vrtadmin/clamav-devel/blob/0.99.1/etc/clamd.conf.sample#L437

**Comment by VicDeo** [comment]
I mean misplaced zero chunk is fixed. The one that gives 
```
got command INSTREAM (9, 14), argument:
Receive thread: INSTREAM: /tmp//clamav-792e5778564386ca029f05178f41a7e2.tmp fd 11
Breaking command loop, mode is no longer MODE_COMMAND
Moved partial command: 680
mode == MODE_STREAM
Got chunksize: 0
Chunks complete
```

On the clamd side.

But
```
"fwrite(): send of 4 bytes failed with errno=32 Broken pipe at \/var\/www\/owncloud\/apps\/files_antivirus\/lib\/scanner.php#68"
```

in owncloud log also logged when the file exceeds clamd limits.

**Comment by michag86** [comment]
Thats true. I solved this with setting MaxFileSize to the same value as maxuploadsize in ownCloud settings. In my case this is 5GB.

**Comment by bboule** [comment]
@felixboehm Looks like the backports are what you needed for this case?  I assume we can move this to resolve (which I did) let me know if otherwise

**Comment by PVince81** [comment]
Any update ? Can this be closed or is there anything left to be fixed ?

**Comment by MorrisJobke** [comment]
This is available in 8.1.5 and 8.2.2 -> closing
